### PR TITLE
Teach client connection about "Sec-WebSocket-Protocol" header.

### PR DESCRIPTION
**Pull request by ewak**
To simplify testing of ui.subprotocol, using just the uWebSockets library, teach the `uWS::WebSocket<uWS::CLIENT>` how to set an optional Sec-WebSocket-Protocol header.

To attempt to minimise API change this is done by adding an extra optional parameter to the end of the Hub::connect() function.

### CODE REVIEW

File: src/HTTPSocket.cpp [hunk 1]
```diff
@@ -246,11 +246,16 @@ bool HttpSocket<isServer>::upgrade(const char *secKey, const char *extensions, s
             return false;
         }
     } else {
+	std::string optionalSubprotocol;
+	if(!getData()->subprotocol.empty()) {
```
Comment (alexhultman) [inline src/HTTPSocket.cpp]: tab / spaces mixed

File: src/HTTPSocket.h [hunk 2]
```diff
@@ -116,6 +116,7 @@ struct WIN32_EXPORT HttpSocket : private uS::Socket {
         // todo: limit these to only client, but who cares for now?
         std::string path;
         std::string host;
+        std::string subprotocol;
```
Comment (alexhultman) [inline src/HTTPSocket.h]: okay

File: src/Hub.cpp [hunk 3]
```diff
@@ -117,8 +117,9 @@ void Hub::connect(std::string uri, void *user, int timeoutMs, Group<CLIENT> *eh)
         httpSocketData->host = hostname;
         httpSocketData->path = path;
         httpSocketData->httpUser = user;
+        httpSocketData->subprotocol = subprotocol;
 
-        uS::Socket s = uS::Node::connect<onClientConnection>(hostname.c_str(), port, secure, httpSocketData);
+        uS::Socket s = uS::Node::connect<onClientConnection>(hostname.c_str(), port, secure, httpSocketData, subprotocol.c_str());
```
Comment (alexhultman) [inline src/Hub.cpp]: subprotocol is not part of the TCP layer, don't pass it here

File: src/Node.h [hunk 4]
```diff
@@ -33,7 +33,7 @@ class WIN32_EXPORT Node {
     }
 
     template <void C(Socket p, bool error)>
-    uS::Socket connect(const char *hostname, int port, bool secure, uS::SocketData *socketData) {
+    uS::Socket connect(const char *hostname, int port, bool secure, uS::SocketData *socketData, const char *subprotocol) {
```
Comment (alexhultman) [inline src/Node.h]: also here, don't change things in the TCP layer

**Comment by alexhultman** [comment]
There are two problems: you shouldn't change anything in the Socket.h / Node.h files, and you have lots of styling issues like mixed spaces and tabs and too many spaces and such. Otherwise it looks fine

**Comment by ewak** [comment]
Tried to fix this up but the branching went wrong and force pushing made a mess.
Am redoing it in a new pr.

**Pull request closed without merging**